File: 0224-basic-calculator/0224-basic-calculator.java
class Solution {
    public int calculate(String s) {
        
        Stack<Integer> stack = new Stack<>();

        int curr = 0, resp = 0, sign = 1;

        for (char ch: s.toCharArray()) {

            if (Character.isDigit(ch)) {

                curr = curr*10 + (int)ch - '0';

            } else if (ch == '+' || ch == '-') {
                resp += curr * sign;
                curr = 0;
                sign = ch == '+' ? 1 : -1;
            } else if (ch == '(') {
                stack.push(resp);
                stack.push(sign);
                resp = 0;
                sign = 1;
            } else if (ch == ')') {

                resp += curr * sign;

                resp *= stack.pop();
                resp += stack.pop();

                curr = 0;
            }

        }

        return resp + curr * sign;

    }
}

File: 0224-basic-calculator/0224-basic-calculator.py
class Solution:
    def calculate(self, s: str) -> int:
        stack = []

        curr_sum, sign, res = 0, 1, 0

        for c in s:

            if c.isdigit():
                curr_sum = curr_sum*10 + int(c)

            elif c =='+' or c == '-':
                res += curr_sum * sign
                sign = -1 if c == '-' else 1
                curr_sum = 0
            
            elif c == '(':
                stack.append(res)
                stack.append(sign)
                res, sign = 0, 1
            elif c == ')':

                res += sign * curr_sum

                res *= stack.pop()
                res += stack.pop()
                curr_sum = 0
                
        return res + curr_sum * sign